### PR TITLE
[AEA-1356] Dialogues subsystem memory optimization

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

Dialogues subsystem memory optimization

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/protocols/dialogue/base.py [hunk 1]
```diff
@@ -507,22 +538,6 @@ def _get_message(self, message_id: int) -> Message:
             raise ValueError("Message not present.")
         return message
 
-    def _has_message(self, message: Message) -> bool:
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:510]: No longer needed?

**Review by DavidMinarsch — CHANGES_REQUESTED** [review]
Current implementation does not allow for extension of the dialogue and dialogues attributes. This leads now to many test failures.

File: aea/protocols/dialogue/base.py (hunk 1 above)
Comment (DavidMinarsch) [reply aea/protocols/dialogue/base.py:510]: yes, was only used in tests

File: aea/protocols/dialogue/base.py [hunk 2]
```diff
@@ -188,7 +188,31 @@ def from_str(cls, obj: str) -> "DialogueLabel":
         return dialogue_label
 
 
-class Dialogue(ABC):
+class _DialogueMeta(type):
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:191]: Open questions:
- Memory footprint of extensions (additional attributes) using slots vs extensions using body (lazy instantiated). There is an obvious LOC saving with the latter approach over the former. (high priority)
- Describe alternative approach to remove additional attributes from dialogue. (low priority)

File: aea/protocols/dialogue/base.py [hunk 3]
```diff
@@ -217,6 +212,9 @@ class _DialogueMeta(type):
 
     def __new__(cls, name: str, bases: Tuple[Type], dct: Dict) -> "_DialogueMeta":
         """Construct a new type."""
+        # apply empty slots if no slots defined.
+        # temporary disable dct["__slots__"] = dct.get("__slots__", tuple())
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:216]: Not sure I understand?

File: aea/protocols/dialogue/base.py [hunk 4]
```diff
@@ -132,11 +131,7 @@ def dialogue_starter_addr(self) -> str:
     def __eq__(self, other: Any) -> bool:
         """Check for equality between two DialogueLabel objects."""
         if isinstance(other, DialogueLabel):
-            return (
-                self.dialogue_reference == other.dialogue_reference
-                and self.dialogue_starter_addr == other.dialogue_starter_addr
-                and self.dialogue_opponent_addr == other.dialogue_opponent_addr
-            )
+            return hash(self) == hash(other)
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:134]: Explain?

**Review by DavidMinarsch — COMMENTED** [review]
__slots__ declaration missing in the following skills:
- coin_price
- confirmation_aw1
- erc1155_client/deploy
- ml_train
- simple_oracle
- simple_oracle_client
- tac_control/contract
- tac_negotiation

File: aea/protocols/dialogue/base.py (hunk 3 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py:216]: to pass some tests with dialogues missing slots

File: aea/protocols/dialogue/base.py (hunk 4 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py:134]: dont remember right now but assume sloted objects will have unique hash over every slot field

Comment (solarw) [reply aea/protocols/dialogue/base.py:134]: anyway need to check, cause implemented long time ago

File: aea/protocols/dialogue/base.py (hunk 3 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py:216]: so will remove after other fixes done

**Comment by solarw** [comment]
> **slots** declaration missing in the following skills:
> 
> * coin_price
> * confirmation_aw1
> * erc1155_client/deploy
> * ml_train
> * simple_oracle
> * simple_oracle_client
> * tac_control/contract
> * tac_negotiation

I know, so slot system partially disabled to check branch in sync with develop, then will fix dialogues left

File: aea/protocols/dialogue/base.py (hunk 4 above)
Comment (DavidMinarsch) [reply aea/protocols/dialogue/base.py:134]: did you check now?

File: aea/test_tools/test_cases.py [hunk 5]
```diff
@@ -73,6 +73,7 @@
 
 DEFAULT_PROCESS_TIMEOUT = 120
 DEFAULT_LAUNCH_TIMEOUT = 10
+ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
```
Comment (DavidMinarsch) [inline aea/test_tools/test_cases.py:76]: This makes no sense! This is not well defined outside the agents-aea repo.

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Review by jrriehl — COMMENTED** [review]
LGTM

**Pull request merged**